Need `cargo doc --all-features` test in CI/CD

Currently there is no checks to ensure the cargo documentation builds. To ensure no code changes break documentation, we need to ensure that `cargo doc --all-features` passes for every crate.

it should be sufficient to run `cargo doc --all-features` under the `protocols` workspace

this is not a comprehensive analysis (maybe more crates are broken), but I tried it locally and it seems `template_distribution_sv2` is broken:

```
    Checking template_distribution_sv2 v1.0.3 (/Users/plebhash/develop/stratum/protocols/v2/subprotocols/template-distribution)
error[E0433]: failed to resolve: use of undeclared type `Vec`
   --> v2/subprotocols/template-distribution/src/new_template.rs:209:28
    |
209 |             let mut path = Vec::new();
    |                            ^^^ use of undeclared type `Vec`
    |
help: consider importing one of these items
    |
7   + use alloc::vec::Vec;
    |
7   + use crate::vec::Vec;
    |

error[E0433]: failed to resolve: use of undeclared type `Vec`
   --> v2/subprotocols/template-distribution/src/new_template.rs:230:34
    |
230 |             coinbase_tx_outputs: Vec::new().try_into().unwrap(),
    |                                  ^^^ use of undeclared type `Vec`
    |
help: consider importing one of these items
    |
7   + use alloc::vec::Vec;
    |
7   + use crate::vec::Vec;
    |

error[E0599]: no method named `try_into` found for struct `Vec<u8>` in the current scope
   --> v2/subprotocols/template-distribution/src/new_template.rs:195:66
    |
195 |         let coinbase_prefix: binary_sv2::B0255 = coinbase_prefix.try_into().unwrap();
    |                                                                  ^^^^^^^^
    |
   ::: /Users/plebhash/.rustup/toolchains/1.78-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/convert/mod.rs:611:8
    |
611 |     fn try_into(self) -> Result<T, Self::Error>;
    |        -------- the method is available for `Vec<u8>` here
    |
    = help: items from traits can only be used if the trait is in scope
    = note: 'core::convert::TryInto' is included in the prelude starting in Edition 2021
help: trait `TryInto` which provides `try_into` is implemented but not in scope; perhaps you want to import it
    |
1   + use core::convert::TryInto;
    |
help: there is a method `into` with a similar name
    |
195 |         let coinbase_prefix: binary_sv2::B0255 = coinbase_prefix.into().unwrap();
    |                                                                  ~~~~

Some errors have detailed explanations: E0433, E0599.
For more information about an error, try `rustc --explain E0433`.
error: could not compile `template_distribution_sv2` (lib) due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
```

this is not a comprehensive analysis (maybe more crates are broken), but I tried it locally and it seems `template_distribution_sv2` is broken:

```
    Checking template_distribution_sv2 v1.0.3 (/Users/plebhash/develop/stratum/protocols/v2/subprotocols/template-distribution)
error[E0433]: failed to resolve: use of undeclared type `Vec`
   --> v2/subprotocols/template-distribution/src/new_template.rs:209:28
    |
209 |             let mut path = Vec::new();
    |                            ^^^ use of undeclared type `Vec`
    |
help: consider importing one of these items
    |
7   + use alloc::vec::Vec;
    |
7   + use crate::vec::Vec;
    |

error[E0433]: failed to resolve: use of undeclared type `Vec`
   --> v2/subprotocols/template-distribution/src/new_template.rs:230:34
    |
230 |             coinbase_tx_outputs: Vec::new().try_into().unwrap(),
    |                                  ^^^ use of undeclared type `Vec`
    |
help: consider importing one of these items
    |
7   + use alloc::vec::Vec;
    |
7   + use crate::vec::Vec;
    |

error[E0599]: no method named `try_into` found for struct `Vec<u8>` in the current scope
   --> v2/subprotocols/template-distribution/src/new_template.rs:195:66
    |
195 |         let coinbase_prefix: binary_sv2::B0255 = coinbase_prefix.try_into().unwrap();
    |                                                                  ^^^^^^^^
    |
   ::: /Users/plebhash/.rustup/toolchains/1.78-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/convert/mod.rs:611:8
    |
611 |     fn try_into(self) -> Result<T, Self::Error>;
    |        -------- the method is available for `Vec<u8>` here
    |
    = help: items from traits can only be used if the trait is in scope
    = note: 'core::convert::TryInto' is included in the prelude starting in Edition 2021
help: trait `TryInto` which provides `try_into` is implemented but not in scope; perhaps you want to import it
    |
1   + use core::convert::TryInto;
    |
help: there is a method `into` with a similar name
    |
195 |         let coinbase_prefix: binary_sv2::B0255 = coinbase_prefix.into().unwrap();
    |                                                                  ~~~~

Some errors have detailed explanations: E0433, E0599.
For more information about an error, try `rustc --explain E0433`.
error: could not compile `template_distribution_sv2` (lib) due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
```

cc @jbesraa

> this is not a comprehensive analysis (maybe more crates are broken), but I tried it locally and it seems `template_distribution_sv2` is broken:
> 
> ```
>     Checking template_distribution_sv2 v1.0.3 (/Users/plebhash/develop/stratum/protocols/v2/subprotocols/template-distribution)
> error[E0433]: failed to resolve: use of undeclared type `Vec`
>    --> v2/subprotocols/template-distribution/src/new_template.rs:209:28
>     |
> 209 |             let mut path = Vec::new();
>     |                            ^^^ use of undeclared type `Vec`
>     |
> help: consider importing one of these items
>     |
> 7   + use alloc::vec::Vec;
>     |
> 7   + use crate::vec::Vec;
>     |
> 
> error[E0433]: failed to resolve: use of undeclared type `Vec`
>    --> v2/subprotocols/template-distribution/src/new_template.rs:230:34
>     |
> 230 |             coinbase_tx_outputs: Vec::new().try_into().unwrap(),
>     |                                  ^^^ use of undeclared type `Vec`
>     |
> help: consider importing one of these items
>     |
> 7   + use alloc::vec::Vec;
>     |
> 7   + use crate::vec::Vec;
>     |
> 
> error[E0599]: no method named `try_into` found for struct `Vec<u8>` in the current scope
>    --> v2/subprotocols/template-distribution/src/new_template.rs:195:66
>     |
> 195 |         let coinbase_prefix: binary_sv2::B0255 = coinbase_prefix.try_into().unwrap();
>     |                                                                  ^^^^^^^^
>     |
>    ::: /Users/plebhash/.rustup/toolchains/1.78-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/convert/mod.rs:611:8
>     |
> 611 |     fn try_into(self) -> Result<T, Self::Error>;
>     |        -------- the method is available for `Vec<u8>` here
>     |
>     = help: items from traits can only be used if the trait is in scope
>     = note: 'core::convert::TryInto' is included in the prelude starting in Edition 2021
> help: trait `TryInto` which provides `try_into` is implemented but not in scope; perhaps you want to import it
>     |
> 1   + use core::convert::TryInto;
>     |
> help: there is a method `into` with a similar name
>     |
> 195 |         let coinbase_prefix: binary_sv2::B0255 = coinbase_prefix.into().unwrap();
>     |                                                                  ~~~~
> 
> Some errors have detailed explanations: E0433, E0599.
> For more information about an error, try `rustc --explain E0433`.
> error: could not compile `template_distribution_sv2` (lib) due to 3 previous errors
> warning: build failed, waiting for other jobs to finish...
> ```
> 
> cc @jbesraa

I dont think there is a problem here. `prop_test` feature requires `std` and we have `no_std` feature. if you run `cargo b --all--features` you enable `prop_test` and `no_std` (and `with_serde`) features, which means `std` is not available but `prop_test` need it

--- a/.github/workflows/docs.yaml
+++ b/.github/workflows/docs.yaml
@@ -1,0 +1,111 @@
+# Enforces Rust Docs sanity (protocols crates only)
+# If `cargo doc --all-features` fails for one crate, the entire workflow fails
+
+name: Rust Docs
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - main
+
+jobs:
+  cargo-doc-all-features:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - uses: actions/checkout@v4
+      - uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+          override: true
+
+      - name: Rust Docs crate common
+        run: |
+          cd common
+          cargo doc
+
+      - name: Rust Docs crate buffer_sv2
+        run: |
+          cd utils/buffer
+          cargo doc
+
+      - name: Rust Docs crate no_serde_sv2_derive_codec
+        run: |
+          cd protocols/v2/binary-sv2/no-serde-sv2/derive_codec
+          cargo doc
+
+      - name: Rust Docs crate no_serde_sv2_codec
+        run: |
+          cd protocols/v2/binary-sv2/no-serde-sv2/codec
+          cargo doc --features with_buffer_pool
+
+      - name: Rust Docs crate serde_sv2
+        run: |
+          cd protocols/v2/binary-sv2/serde-sv2
+          cargo doc
+
+      - name: Rust Docs crate binary_sv2
+        run: |
+          cd protocols/v2/binary-sv2/binary-sv2
+          cargo doc --features core,with_buffer_pool
+
+      - name: Rust Docs crate const_sv2
+        run: |
+          cd protocols/v2/const-sv2
+          cargo doc
+
+      - name: Rust Docs crate framing_sv2
+        run: |
+          cd protocols/v2/framing-sv2
+          cargo doc --features with_buffer_pool
+
+      - name: Rust Docs crate noise_sv2
+        run: |
+          cd protocols/v2/noise-sv2
+          cargo doc --features std
+
+      - name: Rust Docs crate codec_sv2
+        run: |
+          cd protocols/v2/codec-sv2
+          cargo doc --features with_buffer_pool,noise_sv2
+
+      - name: Rust Docs crate common_messages
+        run: |
+          cd protocols/v2/subprotocols/common-messages
+          cargo doc
+
+      - name: Rust Docs crate job_declaration
+        run: |
+          cd protocols/v2/subprotocols/job-declaration
+          cargo doc --all-features
+
+      - name: Rust Docs crate mining
+        run: |
+          cd protocols/v2/subprotocols/mining
+          cargo doc --all-features
+
+      - name: Rust Docs crate template_distribution
+        run: |
+          cd protocols/v2/subprotocols/template-distribution
+          cargo doc
+
+      - name: Rust Docs crate sv2_ffi
+        run: |
+          cd protocols/v2/sv2-ffi
+          cargo doc
+
+      - name: Rust Docs crate roles_logic_sv2
+
+        run: |
+          cd protocols/v2/roles-logic-sv2
+          cargo doc
+
+      - name: Rust Docs crate sv1_api
+        run: |
+          cd protocols/v1
+          cargo doc

